packages/vscode-extension/src/webviews/connectionPanel.ts

```diff
@@ -0,0 +1,251 @@
+      {
+        localResourceRoots: [
+          Uri.joinPath(extensionUri, 'dist', 'webviews'),
+          Uri.joinPath(extensionUri, 'resources'),
+        ],
+        enableScripts: true,
+      },
+    );
+
+    ConnectionPanel._currentPanel = new ConnectionPanel(
+      panel,
+      extensionUri,
+      connection,
+    );
+  }
+
+  private dispose() {
+    ConnectionPanel._currentPanel = undefined;
+
+    this._panel.dispose();
+
+    while (this._disposables.length) {
+      const x = this._disposables.pop();
+      if (x) {
+        x.dispose();
+      }
+    }
+  }
+
+  private update(): void {
+    const webview = this._panel.webview;
+    this._panel.webview.html = this.getHtmlForWebview(webview);
+
+    webview.onDidReceiveMessage(
+      async (message: ConnectionPanelMessage) => {
+        switch (message.command) {
+          case 'onSaveConnection': {
+            await commands.executeCommand(
+              constants.COMMANDS.SAVE_CONNECTION_COMMAND,
+              message.connection,
+              message.password,
+            );
+            this.dispose();
+            break;
+          }
+          case 'onValidationError': {
+            void window.showErrorMessage(
+              constants.MESSAGES.CONNECTION_VALIDATION_MESSAGE,
+            );
+            break;
+          }
+        }
+      },
+      null,
+      this._disposables,
+    );
+  }
+
+  private getHtmlForWebview(webview: Webview): string {
+    const resetCssPath = Uri.joinPath(
+      this._extensionUri,
+      'resources',
+      'styles',
+      'reset.css',
+    );
+    const vscodeCssPath = Uri.joinPath(
+      this._extensionUri,
+      'resources',
+      'styles',
+      'vscode.css',
+    );
+    const connectionPanelCssPath = Uri.joinPath(
+      this._extensionUri,
+      'resources',
+      'styles',
+      'connectionPanel.css',
+    );
+
+    const connectionPanelJsPath = Uri.joinPath(
+      this._extensionUri,
+      'dist',
+      'webviews',
+      'connectionPanelController.js',
+    );
+
+    const resetCssUri = webview.asWebviewUri(resetCssPath);
+    const vscodeCssUri = webview.asWebviewUri(vscodeCssPath);
+    const connectionPanelCssUri = webview.asWebviewUri(connectionPanelCssPath);
+    const connectionPanelJsUri = webview.asWebviewUri(connectionPanelJsPath);
+    const nonce = getNonce();
+
+    // Connect by default if this is a new connection
+    const connect = this._connection?.connect ?? true;
+
+    return `
+        <!DOCTYPE html>
+        <html lang="en">
+          <head>
+            <meta charset="UTF-8">
+            <!--
+            Use a content security policy to only allow loading images from https or from our extension directory,
+            and only allow scripts that have a specific nonce.
+            -->
+            <meta http-equiv="Content-Security-Policy" content="img-src https: data:; style-src 'unsafe-inline' ${
+              webview.cspSource
+            }; script-src 'nonce-${nonce}';">
```

Comment:
we use a nonce to ensure that only scripts with this nonce are allowed to execute inside of our webview